packages/cli/src/wiring/CliInitializer.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import yargs from 'yargs';
+import { CommandGraph } from './CommandGraph';
+import { CliFeature, InternalCliFeature, InternalCliPlugin } from './types';
+import { CommandRegistry } from './CommandRegistry';
+import chalk from 'chalk';
+
+type UninitializedFeature = CliFeature | Promise<CliFeature>;
+
+export class CliInitializer {
+  private graph = new CommandGraph();
+  private commandRegistry = new CommandRegistry(this.graph);
+  #uninitiazedFeatures: Promise<CliFeature>[] = [];
+
+  add(module: UninitializedFeature) {
+    this.#uninitiazedFeatures.push(Promise.resolve(module));
+  }
+
+  async #register(feature: CliFeature) {
+    if (isCliPlugin(feature)) {
+      await feature.init(this.commandRegistry);
+    } else {
+      throw new Error(`Unsupported feature type: ${feature.$$type}`);
+    }
+  }
+
+  async #doInit() {
+    const features = await Promise.all(this.#uninitiazedFeatures);
+    for (const feature of features) {
+      await this.#register(feature);
+    }
+  }
+
+  /**
+   * Actually parse argv and pass it to the command.
+   */
+  async run() {
+    await this.#doInit();
+
+    const {
+      _: commandName,
+      $0: binaryName,
+      ...options
+    } = await yargs(process.argv.slice(2)).parse();
+    const command = this.graph.find(commandName.map(String));
+    if (!command) {
+      console.error(chalk.red(`Command not found: "${commandName.join(' ')}"`));
+      const possibleCommands = this.graph.atDepth(commandName.length - 1);
+      if (possibleCommands.length > 0) {
+        console.log('Available commands:');
+        for (const node of possibleCommands) {
+          let text = `\t${node.name}`;
+          if (node.$$type === '@tree/root') {
+            text += ' [command]';
+          } else {
+            text += ` [options]\t\t${node.command.description}`;
+          }
+          console.log(text);
+        }
+      }
+      process.exit(1);
+    }
+    const parsedOptions = command.schema.parse(options);
```

Comment:
I don't think this is gonna work very well in practice tbh, you kinda need to know what the option structure is when you're parsing the args.

For example, parsing these directly with `yargs` gives you different results:

```
yargs().parse('--fix --config app-config.yaml file.txt')
yargs().parse('--config app-config.yaml --fix file.txt')
```

But if you define `fix` as a flag upfront, you get what you'd arguably want:

```
yargs().option('fix', { type: 'boolean' }).parse('--config app-config.yaml --fix file.txt')
```

I'm hoping we'll be able to lean on existing libraries to handle the arg parsing within plugins, that we don't need to know about or handle that at the framework level at all. So to be clear, I think we should pass an array of all raw args that appear after the command itself in input argv to `command.execute`.

Reply:
Do you have an interface in mind when you say array of raw args? Thinking something like `{optionName: string; type:'boolean' | 'string' | 'number', variadic: boolean }`

Reply:
Plain and simple `string[]`:

```ts
reg.addCommand({
  path: ['config:check'],
  description:
    'Validate that the given configuration loads and matches schema',
  execute: async ({ args }) => {
    const options = yargs()
      .option('lax', { type: 'boolean' })
      /* ... */
      .parse(args);
    const m = await import('./commands/validate');
    await m.default(options);
  },
});
```

Reply:
Ohhhh, gotcha, summarizing for my understanding - each command gets to use whichever arg parser they want, on the framework end we would also use an arg parser (say `yargs`), but they'd be separate instances. Was a little hungup on how to handle help pages, but I guess it wouldn't matter since they'd just pass the `--help` to the command which would parse it and show its own help. Tree-level help would still be provided by the framework.

Reply:
Yep, spot on! 👍